### ISSUE TITLE
DAPO token level loss is incorrectly implemented

### ISSUE THREAD

**Issue by jiosephlee** [huggingface/trl#5375]
https://github.com/huggingface/trl/blob/7f1dd11bc0d0951d0ee6e621caeee70b14df0acf/trl/experimental/gspo_token/grpo_trainer.py#L119

It seems like TRL currently acheives the token level loss proposed by DAPO by treating all tokens equally in the batch. If this is indeed true, this is different from what DAPO actually proposes. DAPO treats all tokens across generations for its respective prompt equally, rather than treating all tokens in a given batch (which could include multiple prompts and its generations) equally.

<img width="888" height="178" alt="Image" src="https://github.com/user-attachments/assets/7e90e83b-22f8-421f-b0e9-7ef1fe848b44" />

Granted, the paper doesn't make it explicitly clear whether prompts should then be equally averaged or further balanced on a token-level, it seems like follow-up papers such as [ScaleRL](https://arxiv.org/pdf/2510.13786) presume DAPO means token-level aggregation at the prompt level.

<img width="971" height="142" alt="Image" src="https://github.com/user-attachments/assets/0e50f27d-bcfb-4d4d-9f20-33ca18923c47" />

**Comment by qgallouedec** [comment]
Thank for opening the issue. I think that the default `loss_agg_mode` in verl's DAPO (reference implementation for DAPO) is `"token-mean"`, see [here](https://github.com/volcengine/verl/blob/51da8fea41da1244a0cadaa39de03dc06127ef4b/verl/trainer/config/actor/actor.yaml#L82) and [here](https://github.com/verl-project/verl/blob/7402ca73bcf2b85b5337393b4ccc9ec45ea96b6d/verl/trainer/ppo/core_algos.py#L1173)

```python
loss = verl_F.masked_sum(loss_mat, loss_mask) / batch_num_tokens * dp_size
```

where `batch_num_tokens` is the global count of valid response tokens (all-reduced across DP ranks, see [here](https://github.com/volcengine/verl/blob/51da8fea41da1244a0cadaa39de03dc06127ef4b/verl/workers/engine/fsdp/transformer_impl.py#L579-L583)). This is a flat token average (every token across every response and every prompt contributes equally). Consequently, prompts whose responses are collectively longer have proportionally more influence on the loss.

This is not prompt-average. Prompt-average would require first averaging the $G$ losses within each prompt's response group, then averaging across prompts. AFAICT, none of the available `loss_agg_mode` options do this, (even `seq-mean-token-sum` averages over individual sequences, not prompts).


It might be an error in the ScaleRL paper

**Comment by qgallouedec** [comment]
tagging @wuxibin89 if they can confirm 🙏

**Comment by jiosephlee** [comment]
I see; I didn't realize the DAPO author's reference implementation was in `verl`. I personally actively work with OpenRLHF and it also has a different (wrong) implementation of DAPO's token-level aggregation, so I've been wary of current implementations. 

Taxonomy-wise, I do think prompt level aggregation fits in naturally between sequence-level token loss aggregation and batch-level token loss aggregation. I'm not sure if any other papers besides ScaleRL examines this but it could be worth supporting (reportedly works the best in their ablation).

**Comment by qgallouedec** [comment]
we could consider an additional loss_type for it indeed. We just need to find a good name.

**Comment by FiorenzoParascandolo1** [comment]
I think there might be an ambiguity here regarding how the normalization should behave when the generation batch is split across multiple optimizer steps.

From my understanding, the current implementation effectively assumes a global token normalization (i.e. equivalent to averaging over all tokens in the full generation batch). However, this is only strictly correct if all microbatches contribute to a single optimizer step.

In practice, if we instead:

generate a batch of size N (e.g. 1024 samples, 3000 active tokens total),
split it into K training batches (e.g. 2 batches of 512),
and perform K separate optimizer.step() calls,

then each step is only seeing a subset of the tokens.

For example:

Generation batch: 3000 active tokens
Split into:
batch 1: 1000 tokens
batch 2: 2000 tokens

If we do two optimizer steps, shouldn't the normalization be:

step 1 → divide by 1000
step 2 → divide by 2000

instead of dividing both by 3000?

Using the global token count (3000) for each step would only be correct if both batches contribute to the same accumulated update. Otherwise, it effectively underweights each step.

So the question is: is num_items_in_batch intended to represent the local token count for the current optimizer step, or the global token count across the full generation batch?

**Comment by qgallouedec** [comment]
Just to close the loop on the original bug claim: TRL's `dapo` matches verl's reference (`loss_agg_mode="token-mean"` = flat token average across all responses and all prompts in the batch), so this isn't a deviation from DAPO as authored.

The prompt-level aggregation variant from _ScaleRL_ is a different recipe, and a reasonable thing to support eventually. But I'd like to see a bit more signal (other users running into it, or a draft PR) before adding another `loss_type` to the public config surface. Happy to review a draft PR if anyone watching wants to take a shot; otherwise let's leave this open as a tracking thread.


@FiorenzoParascandolo1, your accumulation question is real but tangential to this thread; it could be worth opening a separate issue so it doesn't get lost.

**Comment by DagaBhai** [comment]
> Just to close the loop on the original bug claim: TRL's `dapo` matches verl's reference (`loss_agg_mode="token-mean"` = flat token average across all responses and all prompts in the batch), so this isn't a deviation from DAPO as authored.
> 
> The prompt-level aggregation variant from _ScaleRL_ is a different recipe, and a reasonable thing to support eventually. But I'd like to see a bit more signal (other users running into it, or a draft PR) before adding another `loss_type` to the public config surface. Happy to review a draft PR if anyone watching wants to take a shot; otherwise let's leave this open as a tracking thread.
> 
> [@FiorenzoParascandolo1](https://github.com/FiorenzoParascandolo1), your accumulation question is real but tangential to this thread; it could be worth opening a separate issue so it doesn't get lost.

I have opened this PR based on my understanding of the issue. let me know if any changes are required

**Comment by qgallouedec** [comment]
Hey @jiosephlee, sorry for the slow follow-up. Good news: no new loss_type is actually needed here.

We already confirmed `dapo` matches DAPO's own reference implementation (verl's flat token-mean), so that part isn't a bug.

The prompt-level alternative you and ScaleRL describe (average within each prompt's group of completions, then average across prompts) turns out to be, i think, what `loss_type="grpo"` already computes in TRL today. Every prompt always contributes the same number of completions (`num_generations`), and for equal-size groups, "average of group-means" and "flat average over all completions" are the same number, arithmetically:

https://github.com/huggingface/trl/blob/6d484baed4db18c2799b9d5b32c2520cba6b4df2/trl/trainer/grpo_trainer.py#L3242-L3246

That's why both #5381 and #5828 (a second attempt at a `dapo_prompt_mean` loss_type) reduce to a no-op relative to `grpo`, just verified it numerically.

So: 
want DAPO's paper weighting (all tokens in the batch equal)? Use `dapo`
want prompt-level weighting instead? `grpo` already gives you that, no new option needed.

Closing this out, but reopen if you find a case where they actually diverge.
